File: src/domain/theme.ts
export type AppThemeId =
  | "vscode-dark"
  | "vscode-light"
  | "github-dark"
  | "github-light"
  | "dracula"
  | "night-owl"
  | "monokai"
  | "paper";

export type AppThemeOption = {
  readonly id: AppThemeId;
  readonly label: string;
};

export type ThemePreferences = {
  readonly appThemeId: AppThemeId;
  readonly previewThemeId: string | null;
  readonly previewUsesAppThemeColors: boolean;
};

export const APP_THEME_OPTIONS: readonly AppThemeOption[] = [
  { id: "vscode-dark", label: "VS Code Dark" },
  { id: "vscode-light", label: "VS Code Light" },
  { id: "github-dark", label: "GitHub Dark" },
  { id: "github-light", label: "GitHub Light" },
  { id: "dracula", label: "Dracula" },
  { id: "night-owl", label: "Night Owl" },
  { id: "monokai", label: "Monokai" },
  { id: "paper", label: "Paper" },
] as const;

const APP_THEME_ID_SET = new Set<AppThemeId>(APP_THEME_OPTIONS.map((themeOption) => themeOption.id));

export function isAppThemeId(value: string): value is AppThemeId {
  return APP_THEME_ID_SET.has(value as AppThemeId);
}

File: crates/kmark-core/src/theme.rs
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum AppThemeId {
    VscodeDark,
    VscodeLight,
    GithubDark,
    GithubLight,
    Dracula,
    NightOwl,
    Monokai,
    Paper,
}

impl Default for AppThemeId {
    fn default() -> Self {
        Self::VscodeDark
    }
}

impl AppThemeId {
    pub fn from_str(value: &str) -> Option<Self> {
        match value {
            "vscode-dark" => Some(Self::VscodeDark),
            "vscode-light" => Some(Self::VscodeLight),
            "github-dark" => Some(Self::GithubDark),
            "github-light" => Some(Self::GithubLight),
            "dracula" => Some(Self::Dracula),
            "night-owl" => Some(Self::NightOwl),
            "monokai" => Some(Self::Monokai),
            "paper" => Some(Self::Paper),
            _ => None,
        }
    }

    pub fn as_str(self) -> &'static str {
        match self {
            Self::VscodeDark => "vscode-dark",
            Self::VscodeLight => "vscode-light",
            Self::GithubDark => "github-dark",
            Self::GithubLight => "github-light",
            Self::Dracula => "dracula",
            Self::NightOwl => "night-owl",
            Self::Monokai => "monokai",
            Self::Paper => "paper",
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ThemePreferences {
    app_theme_id: AppThemeId,
    preview_theme_id: Option<String>,
    preview_uses_app_theme_colors: bool,
}

impl Default for ThemePreferences {
    fn default() -> Self {
        Self {
            app_theme_id: AppThemeId::VscodeDark,
            preview_theme_id: None,
            preview_uses_app_theme_colors: false,
        }
    }
}

impl ThemePreferences {
    pub fn new(
        app_theme_id: Option<&str>,
        preview_theme_id: Option<&str>,
        preview_uses_app_theme_colors: Option<bool>,
    ) -> Self {
        let defaults = Self::default();

        Self {
            app_theme_id: app_theme_id
                .and_then(AppThemeId::from_str)
                .unwrap_or(defaults.app_theme_id),
            preview_theme_id: preview_theme_id
                .map(str::trim)
                .filter(|value| !value.is_empty())
                .map(ToOwned::to_owned),
            preview_uses_app_theme_colors: preview_uses_app_theme_colors
                .unwrap_or(defaults.preview_uses_app_theme_colors),
        }
    }

    pub fn app_theme_id(&self) -> AppThemeId {
        self.app_theme_id
    }

    pub fn preview_theme_id(&self) -> Option<&str> {
        self.preview_theme_id.as_deref()
    }

    pub fn preview_uses_app_theme_colors(&self) -> bool {
        self.preview_uses_app_theme_colors
    }
}
